libsolidity/ast/TypeProvider.h

```diff
@@ -0,0 +1,228 @@
+	///          This function is mostly useful to modify inner types appropriately.
+	static Type const* withLocationIfReference(DataLocation _location, Type const* _type)
+	{
+		if (auto refType = dynamic_cast<ReferenceType const*>(_type))
+			return withLocation(refType, _location, false);
+
+		return _type;
+	}
+
+	/// @returns the type of a function.
+	static FunctionType const* functionType(FunctionDefinition const& _function, bool _isInternal = true);
+
+	/// @returns the accessor function type of a state variable.
+	static FunctionType const* functionType(VariableDeclaration const& _varDecl);
+
+	/// @returns the function type of an event.
+	static FunctionType const* functionType(EventDefinition const& _event);
+
+	/// @returns the type of a function type name.
+	static FunctionType const* functionType(FunctionTypeName const& _typeName);
+
+	/// @returns the function type to be used for a plain type (not derived from a declaration).
+	static FunctionType const* functionType(
+		strings const& _parameterTypes,
+		strings const& _returnParameterTypes,
+		FunctionType::Kind _kind = FunctionType::Kind::Internal,
+		bool _arbitraryParameters = false,
+		StateMutability _stateMutability = StateMutability::NonPayable
+	);
+
+	/// @returns a highly customized FunctionType, use with care.
+	static FunctionType const* functionType(
+		TypePointers const& _parameterTypes,
+		TypePointers const& _returnParameterTypes,
+		strings _parameterNames = strings{},
+		strings _returnParameterNames = strings{},
+		FunctionType::Kind _kind = FunctionType::Kind::Internal,
+		bool _arbitraryParameters = false,
+		StateMutability _stateMutability = StateMutability::NonPayable,
+		Declaration const* _declaration = nullptr,
+		bool _gasSet = false,
+		bool _valueSet = false,
+		bool _bound = false
+	);
+
+	/// Auto-detect the proper type for a literal. @returns an empty pointer if the literal does
+	/// not fit any type.
+	static TypePointer forLiteral(Literal const& _literal);
+	static RationalNumberType const* rationalNumberType(Literal const& _literal);
+
+	static RationalNumberType const* rationalNumberType(
+		rational const& _value,
+		Type const* _compatibleBytesType = nullptr
+	);
+
+	static ContractType const* contractType(ContractDefinition const& _contract, bool _isSuper = false);
+
+	static InaccessibleDynamicType const* inaccessibleDynamicType() noexcept { return &m_inaccessibleDynamicType; }
+
+	/// @returns the type of an enum instance for given definition, there is one distinct type per enum definition.
+	static EnumType const* enumType(EnumDefinition const& _enum);
+
+	/// @returns special type for imported modules. These mainly give access to their scope via members.
+	static ModuleType const* moduleType(SourceUnit const& _source);
+
+	static TypeType const* typeType(Type const* _actualType);
+
+	static StructType const* structType(StructDefinition const& _struct, DataLocation _location = DataLocation::Storage);
```

Comment:
We should probably also remove the default value for the location. Also, why is location the first parameter for arrays but the second for structs?